### PR TITLE
Added a deep merger, able to merge collections and beyond.

### PR DESCRIPTION
**Pull request by HeikoStudt**
Added the deep merger implemented for our projects.
Added IMergeableConfigurationCollection interface supporting the DeepMerger

The merger uses the first occurance on attribute level.
Collections are merged via add/remove/clear where several adds with the same key are merged on attribute level.

### CODE REVIEW

**Comment by feitzi** [comment]
please accept this PR!

File: NConfig/Implementation/Mergers/DeepMerger.cs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        {
+            if (sections == null || !sections.Any())
+            {
+                throw new ConfigurationErrorsException("No sections to merge");
+            }
+
+            Type sectionType = sections.First().GetType();
+            ConfigurationSection mergedSection = (ConfigurationSection) Activator.CreateInstance(sectionType);
+
+            new DeepSectionMergerImpl().DoMerge(sections, mergedSection);
```
Comment (Yegoroff) [inline NConfig/Implementation/Mergers/DeepMerger.cs]: I'm not sure why you are use instance here, DeepSectionMergerImpl looks like static class (class without a state), from the first glance.

File: NConfig/Implementation/Mergers/IMergeableConfigurationCollection.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+using System.Collections;
+using System.Configuration;
+
+namespace NConfig
+{
+    /// <summary>
+    ///   Mergeable configuration collection (still untyped).
+    ///   Is an IEnumerable(ConfigurationElement)
+    /// </summary>
+    public interface IMergeableConfigurationCollection : IEnumerable {
```
Comment (Yegoroff) [inline NConfig/Implementation/Mergers/IMergeableConfigurationCollection.cs]: I assume this interface should be implemented by user configuration code. If this is the case, this file shouldn't be inside "Implementation" folder, consider root, next to the other public interfaces.

File: NConfig/Implementation/Mergers/DeepMerger.cs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        public override TSection Merge(IEnumerable<TSection> sections)
+        {
+            if (sections == null || !sections.Any())
+            {
+                throw new ConfigurationErrorsException("No sections to merge");
+            }
+
+            var mergedSection = new TSection();
+
+            new DeepSectionMergerImpl().DoMerge(sections, mergedSection);
```
Comment (Yegoroff) [inline NConfig/Implementation/Mergers/DeepMerger.cs]: why do we need instance of DeepSectionMergerImpl here?

**Comment by HeikoStudt** [comment]
Hi Yegoroff,

thank you for your kind review. I've put into the second commit the following changes including samples.

a) DeepSectionMergerImpl does not use any instance state, therefore it is made to static
b) Added "MergeableConfigurationCollection" base class
c) Added merger section into README.md and some further additions
d) Added Sample for the deep merger


MFG (Kind regards)
Heiko

**Pull request merged**